### PR TITLE
Continue to implement MasterMetadataService to include chunk allocation, and add caching of server locations on client side

### PR DESCRIPTION
**Pull request by xicheng87**
In this PR, we continue to shape up MasterMetadataService to include chunk allocation. More specifically, in the handling of file creation, the master metadata service calls chunk server manager to allocate chunks replica for the created chunk, and adds the chunk server location to the OpenFileReply message.

We also add implementations on the clientside to cache the file chunk metadata once called the file creation rpc. 

Some related data structure tweaks:

1) Add back the repeated field of locations in the FileChunkMetadata message. Although chunk server manager module manages chunk server locations, we need to include these locations in a rpc message so that client can cache them. 

2) Made MasterMetadataServiceImpl::chunk_server_service_clients_ a thread-safe data structure as the server code is intrinsically multi-threaded. 

We'd have to implement more services to test the full functionality of this change, as currently AllocateChunkServer returns an empty set.

### CODE REVIEW

File: src/server/master_server/master_metadata_service_impl.cc [hunk 1]
```diff
@@ -19,13 +26,90 @@ inline server::MetadataManager* MasterMetadataServiceImpl::metadata_manager() {
...
+  // Step 2. Create the first file chunk for this file
+  google::protobuf::util::StatusOr<std::string> chunk_or(
+      metadata_manager()->CreateChunkHandle(filename,0));
+  if (!chunk_or.ok()) {
+    return common::utils::ConvertProtobufStatusToGrpcStatus(status);
+  }
+  const std::string& chunk(chunk_or.ValueOrDie());
+
+  // Step 3. Allocate chunk servers
+  const ushort num_of_chunk_replica(3);
```
Comment (micheal-o) [inline src/server/master_server/master_metadata_service_impl.cc:54]: nit: The default is 3. But i thought client can also config replication count in their request. This is fine for now.

File: src/server/master_server/master_metadata_service_impl.cc [hunk 2]
```diff
@@ -19,13 +26,90 @@ inline server::MetadataManager* MasterMetadataServiceImpl::metadata_manager() {
...
+  // TODO(someone): make this number configurable via config.yml. Probably not
+  // a priority right now
+  auto chunk_server_locations(
```
Comment (micheal-o) [inline src/server/master_server/master_metadata_service_impl.cc:57]: nit: this can be empty, if no location is available. We should log and return fail for this.

File: src/common/protocol_client/grpc_client_utils.cc [hunk 3]
```diff
@@ -0,0 +1,15 @@
+#include "absl/time/time.h"
+#include "src/common/protocol_client/grpc_client_utils.h"
+
```
Comment (Michael-Tu) [inline src/common/protocol_client/grpc_client_utils.cc:3]: add `#include <chrono>`

File: src/server/master_server/master_metadata_service_impl.h [hunk 4]
```diff
@@ -14,16 +17,22 @@ namespace service {
 class MasterMetadataServiceImpl final
     : public protos::grpc::MasterMetadataService::Service {
  public:
+  MasterMetadataServiceImpl(common::ConfigManager* config_manager) :
```
Comment (micheal-o) [inline src/server/master_server/master_metadata_service_impl.h:20]: nit: why this? you can get config manager by doing ConfigManager::GetConfig(). It's singleton. Am I missing something here?

File: src/server/master_server/master_metadata_service_impl.h [hunk 5]
```diff
@@ -14,16 +17,22 @@ namespace service {
...
   bool RegisterChunkServerRpcClient(std::string server_name,
                                     std::shared_ptr<grpc::Channel> channel);
 
-protected:
+ protected:
   // Accessor to the MetadataManager instance
   server::MetadataManager* metadata_manager();
 
+  // Accessor to the ChunkServerManager instance
+  server::ChunkServerManager& chunk_server_manager();
```
Comment (micheal-o) [inline src/server/master_server/master_metadata_service_impl.h:34]: nit: ditto. You can get this directly too (its a singleton), instead of having methods. Reason for this?

File: src/server/master_server/master_metadata_service_impl.cc [hunk 6]
```diff
@@ -19,13 +26,90 @@ inline server::MetadataManager* MasterMetadataServiceImpl::metadata_manager() {
...
+      chunk_server_manager().AllocateChunkServer(chunk, num_of_chunk_replica));
+
+  // Prepare the InitChunkFileReply
+  *reply->mutable_metadata() = FileChunkMetadata();
+  reply->mutable_metadata()->set_chunk_handle(chunk);
+  reply->mutable_metadata()->set_version(0);
+
+  // Step 4. Coordinate with chunk servers to initialize the file chunk 
+  for(auto chunk_server_location : chunk_server_locations) {
+    const std::string server_address(chunk_server_location.server_hostname() +
```
Comment (Michael-Tu) [inline src/server/master_server/master_metadata_service_impl.cc:67]: server address is `hostname:port`, you forgot the colon

File: src/server/master_server/master_metadata_service_impl.cc [hunk 7]
```diff
@@ -19,13 +26,90 @@ inline server::MetadataManager* MasterMetadataServiceImpl::metadata_manager() {
...
+        std::to_string(chunk_server_location.server_port()));
+    auto chunk_server_channel(grpc::CreateChannel(server_address,
+        grpc::InsecureChannelCredentials()));
+    // Register this chunk server Rpc client if not existed
+    RegisterChunkServerRpcClient(server_address, chunk_server_channel);
```
Comment (Michael-Tu) [inline src/server/master_server/master_metadata_service_impl.cc:72]: yikes. in my PR, I deprecated this function to only support `server_address`. prob gonna have to resolve this merge conflict at some point

File: src/server/master_server/master_metadata_service_impl.cc [hunk 8]
```diff
@@ -19,13 +26,90 @@ inline server::MetadataManager* MasterMetadataServiceImpl::metadata_manager() {
...
+    return common::utils::ConvertProtobufStatusToGrpcStatus(status);
+  }
+
+  // Step 2. Create the first file chunk for this file
+  google::protobuf::util::StatusOr<std::string> chunk_or(
+      metadata_manager()->CreateChunkHandle(filename,0));
+  if (!chunk_or.ok()) {
+    return common::utils::ConvertProtobufStatusToGrpcStatus(status);
+  }
+  const std::string& chunk(chunk_or.ValueOrDie());
```
Comment (Michael-Tu) [inline src/server/master_server/master_metadata_service_impl.cc:51]: nit. it'd be really readable if you call this `chunk_handle`; i kept thinking this is file chunk proto or something when reading code

File: src/server/master_server/master_metadata_service_impl.cc [hunk 9]
```diff
@@ -19,13 +26,90 @@ inline server::MetadataManager* MasterMetadataServiceImpl::metadata_manager() {
...
+    // Issue InitFileChunk request and check status
+    google::protobuf::util::StatusOr<InitFileChunkReply> init_chunk_or(
+        chunk_server_service_client->SendRequest(init_chunk_request, 
+            client_context));
+
+    // TODO(Xi): we probably don't have to fail this request just because one
+    // init chunk request is not successful. Will come back and refine this 
+    // once the chunk server grpc impl is in shape
+    if (!init_chunk_or.ok()) {
+      common::utils::ConvertProtobufStatusToGrpcStatus(status);
```
Comment (Michael-Tu) [inline src/server/master_server/master_metadata_service_impl.cc:97]: `return`?

File: src/server/master_server/master_metadata_service_impl.h (hunk 5 above)
Comment (Michael-Tu) [inline src/server/master_server/master_metadata_service_impl.h:34]: `server::ChunkServerManager* chunk_server_manager();` ?

File: src/server/master_server/master_metadata_service_impl.h (hunk 4 above)
Comment (Michael-Tu) [reply src/server/master_server/master_metadata_service_impl.h:20]: Actually I wrote ConfigManager NOT as a singleton; this is just a static method; each time you call GetConfig it parses a new file again

File: src/server/master_server/master_metadata_service_impl.cc (hunk 1 above)
Comment (xicheng87) [reply src/server/master_server/master_metadata_service_impl.cc:54]: Sounds good.

File: src/server/master_server/master_metadata_service_impl.cc (hunk 2 above)
Comment (xicheng87) [reply src/server/master_server/master_metadata_service_impl.cc:57]: Yes, I think for file creation it may not qualify an error because the client may call openFileRequest again the next time, but yes let's keep this in mind.

File: src/server/master_server/master_metadata_service_impl.h (hunk 4 above)
Comment (xicheng87) [reply src/server/master_server/master_metadata_service_impl.h:20]: Thanks for clarifying Gan, correct, because GetConfig is not a singleton (as it takes a file path as input), we need a reference of config manager here.

File: src/server/master_server/master_metadata_service_impl.h (hunk 5 above)
Comment (xicheng87) [reply src/server/master_server/master_metadata_service_impl.h:34]: Writing "server::ChunkServerManager::GetInstance()" and "server::MetadataManager::GetInstance()" takes quite a bit of space, and I am basically using these methods as alias :) Furthermore, they are inlined (see cc file) so no extra overhead.

File: src/common/protocol_client/grpc_client_utils.cc (hunk 3 above)
Comment (xicheng87) [reply src/common/protocol_client/grpc_client_utils.cc:3]: Done

File: src/server/master_server/master_metadata_service_impl.cc (hunk 6 above)
Comment (xicheng87) [reply src/server/master_server/master_metadata_service_impl.cc:67]: Good catch, thanks.

File: src/server/master_server/master_metadata_service_impl.cc (hunk 7 above)
Comment (xicheng87) [reply src/server/master_server/master_metadata_service_impl.cc:72]: Yeah, I think can I understand why you want to support server_address only. If the server_address already exists, we should not need to create the channel. But this becomes a tricky and challenging thing to do with parallel hash map due to the extra operation that is involved, so I have to keep the channel variable.

File: src/server/master_server/master_metadata_service_impl.cc (hunk 8 above)
Comment (xicheng87) [reply src/server/master_server/master_metadata_service_impl.cc:51]: Thanks for pointing this out, I revised it.

File: src/server/master_server/master_metadata_service_impl.h (hunk 5 above)
Comment (xicheng87) [reply src/server/master_server/master_metadata_service_impl.h:34]: good catch.

File: src/server/master_server/master_metadata_service_impl.cc (hunk 9 above)
Comment (xicheng87) [reply src/server/master_server/master_metadata_service_impl.cc:97]: Good catch.

File: src/server/master_server/master_metadata_service_impl.cc (hunk 7 above)
Comment (Michael-Tu) [reply src/server/master_server/master_metadata_service_impl.cc:72]: not quite sure about what you meant by parallel hashmap? FYI, my new way of doing it is instead of having <server_name, client_instance>, i have < server_address, cient_instance> so we can also add more servers in that were not in the initial config file
